### DISCUSSION TITLE
Include AI code generation

### DISCUSSION THREAD

**Discussion by Zorakie** [arnoudkooi/SN-Utils#452] [Ideas]
Hi!
As I was playing with Bing and ChatGPT, I was wondering, if SNUtils could bring AI to the table.

How this should work:

In any code editor, (BRs, SIs, but even background scripts, and even maybe some XMLs or even rich HTML fields), it would be possible to rigth-click and select "Generate here". A pop-up would be shown with the prompt - basically just a multi-line input. There would be some instructions included like "You won't need to write the request to write a ServiceNow code, just state, what the code should do".

Then, on confirmation, SNUtils would wrap the statement into something like
"Write a ServiceNow code."
{prompt from input}
"Provide the code only DO NOT explain the code, DO NOT add confirmation statement, provide the code snippet only."

That seems to be working nicely with ChatGPT 3.5.

![image](https://github.com/arnoudkooi/ServiceNow-Utils/assets/14437576/f45554bd-1a7f-4c74-a381-d1771fd098ea)

Then - because it is still AI, some sanitization would be needed to check, if AI returned only the code snippet, or if we need to cut some generic garbage around.

Then, just parse the code and insert it in the correct place.

Now, there are 2 things that may be tricky.
1) Cross-origin requests, not sure if SNUTILS would be able to call some external API
2) API payments - here, I have seen ChatGPT has APIs behind the firewall, but maybe, just maybe, there could be some workaround through Bing, if that has some free API, as Bing even has GPT-4 now, or there is a second option: 
- SNUtils would open a hidden iframe.
- The user would need to provide his own account (snutils would be able to check, whether the iframe has some content related to the login request, and in that case, the iframe would be shown, so the user can log in, and then immediately hidden again.
- SNUtils would inject the prompt, hit the button, wait for the response, parse the response, and close the iframe.

It is a long-shot, I know that this is not exactly the scope of this application, but who knows :)

**Comment by arnoudkooi** [comment]
I dont want to add a ChatGPT integration.
You could consider using sn-scriptsync to use VS Code and from there work with any AI Plugin available.

**Comment by jacebenson** [comment]
hey @Zorakie I like the idea but the implications of it here I think are just ... well it would probably get the extension removed from places that have to whitelist the extension.  if you want to do something like this.  I have an extension that does this.  If you'd like to contribute, I'd be open to discuss.  You can see that extension on https://scribe.monster  

@arnoudkooi if this is something I should drop from here let me know and I'll readily delete this comment.

**Reply by arnoudkooi** [reply]
Thanks,  good addition, and always welcome to chime in :)

**Reply by Zorakie** [reply]
Hey Jace! Yes, this is basically what I had in my mind, but to be frank, I see some red flags (even if I think your project is probably honest). The first thing is, that it is not open-source, both plugin and whatever server infrastructure. Another middleman, another layer that you must trust to not misuse your data. Moreover, there are APIs for Bing and ChatGPT (e.g. https://github.com/transitive-bullshit/bing-chat) where I do believe it could be possible either to embed the API to the plugin itself, or at least host the server part locally, rather than rely on 3rd party service that plays the role of mailman. For those reasons, unless the project is open-source, I'll pass. But hey, if handled properly, your product could have a great impact in ServiceNow field one day!

**Reply by jacebenson** [reply]
I'd encourage you to make a project that fits your needs.  Chrome extension development is interesting and all that but it's more involved than I thought.  

That project is doing more than just passing the call to an LLM.  Depending on the table/field it's also prepending a number of examples, so it has a great start to give a great answer.  That work I wasn't willing to give away.

Regarding the APIs unless you're running a LLM locally, you'll need to call out to some host you're paying for by token or by time you're hosting.  

In any case like I started with.  I'd love to see someone else's take on this, if you make it, please let me know!
